### PR TITLE
Libvirt e2e CI for pull requests

### LINKED ISSUE #599: Implement the e2e CI for libvirt
As a developer that opened an pull request to confidential-containers/cloud-api-adaptor, I want the CI to run end-to-end tests so that I am sure I didn't introduce regression bugs.

In order to have an e2e CI for libvirt, overall, we will need to:

- [x] Enable libvirt provision on the [e2e test framework](#592)
- [x] Figure out a CI runner. There are a couple of options that we should investigate: Github self-hosted runner, or managed runner with nested, or RH internal host. Related to https://github.com/confidential-containers/community/issues/67
- [ ] Implement the CI itself. E.g. if using Github Actions then implement the workflow yaml

### PR DESCRIPTION
**Pull request by wainersm**
This is my second tentative to have the libvirt e2e tests running for pull requests.

From the user point of view, the workflow is executed by adding the `test_e2e_libvirt` label. In theory if we add support for other providers, it would be a matter of adding a new label. However it may not work as the common jobs (see below) will be triggered every time a `test_e2e_*` label is added to the PR. I prefer to defer that discussion for when (and if) we add other e2e jobs.

Still regarding how it is triggered, an alternative would be to use comments on PR (e.g. `/test_e2e_libvirt`) but then we hit a bunch of limitations on Github actions design. To mention the most basic: on commenting the workflow runs on context of the `main` branch meaning that it doesn't give the reference to the pull request number (or other information like the source tree/branch). That's why I decided to try with the label approach.

Like in the [previous tentative](https://github.com/confidential-containers/cloud-api-adaptor/pull/611) the workflow is split two set of jobs: a) common/shareable among cloud e2e, b) cloud specific e2e  

The common jobs:

 * job1: podvm builder job
 * job2:  image builder job

The **job1** is actually not building podvm as this is a time/resources consuming task. Instead, it is using the `./podvm/hack/download-image.sh` to extract the qcow2 file from already built `quay.io/confidential-containers/podvm-*` images. The qcow2 files are attached to the workflow so that downstream jobs (i.e. the jobs that effectively run e2e tests) can simply download-and-use the disk. Note: I am not sure about this approach of archiving the disk file as they are huge...we may hit storage issues.

The **job2** is going to build and push the cloud-api-adaptor to `ghcr.io/confidential-containers/cloud-api-adaptor`. The kustomization files (`install` dir) are changed to use that new image on installation. So just like in **job1** the entire `install` directory is attached to the workflow so that downstream jobs can simply use it (here the files are a couple of kb so we won't have storage problem).

On more comment on *job2*: as it creates temporary images to `ghcr.io` I implement a script that will clean up unused images. I will watch how this PR evolves to publish its code.

Fixes #599

### CODE REVIEW

**Comment by wainersm** [comment]
Notice that I ran the e2e libvirt workflow by tagging this PR with `test_e2e_libvirt`.

The `e2e tests / image / build and push (dev, linux/amd64)` and `e2e tests / image / build and push (release, linux/amd64,linux/s390x,linux/ppc64le)` jobs failed to push the built CAA image, the reason being that the `pull_request` event doesn't give a `GITHUB_TOKEN` with admin privileges to the workflow (for security reasons).  In order to make it work, we will need to run the entire workflow as `pull_request_target` or the publishing job as `workflow_run` (because those two events will have granted a read/write token for ghcr.io)

Regardless, I would like to receive a feedback on this version first. :)

**Comment by wainersm** [comment]
Updated with a fix to `Install tests dependencies` step on the `libvirt` job. I had added a commit to disable the image build for the sake of testing the `libvirt` job, before the merge I will remove that commit.

**Comment by wainersm** [comment]
Update the PR with:
 * now pinning the version of k8s installed (v1.26.7)
 * downloading the correct version of Ubuntu (20.04 instead of 22.04)

File: .github/workflows/e2e_libvirt.yaml
Comment (stevenhorsman) [inline .github/workflows/e2e_libvirt.yaml:52]: It might not be relevant on the azure version of ubuntu, but just to note that I had to enable the universe repo to install `libvirt-dev` (which I don't remember needing last year?)

**Comment by wainersm** [comment]
Fix https://github.com/confidential-containers/cloud-api-adaptor/pull/1332 is merged, rebased and giving it a try again.

File: .github/workflows/e2e_libvirt.yaml
Comment (wainersm) [reply .github/workflows/e2e_libvirt.yaml:53]: hi @stevenhorsman , the runner is a Ubuntu 22.04 (`Canonical:0001-com-ubuntu-server-jammy:22_04-lts-gen2:latest` image) and it doesn't need to enable such as repo.

**Comment by stevenhorsman** [comment]
> Fix #1332 is merged, rebased and giving it a try again.

FYI the CAA image build failed for this: https://github.com/confidential-containers/cloud-api-adaptor/actions/runs/5939073597/job/16104756661 so it probably won't work yet. Taking a look at that now

**Comment by wainersm** [comment]
re-run the libvirt e2e because the previous execution used an old caa image.

**Comment by wainersm** [comment]
ok, now it is failing to wait for caa install:
```
Wait for the pod cloud-api-adaptor-daemonset-4hbgq be ready
F0822 18:44:38.436873   22803 env.go:369] Setup failure: timed out waiting for the condition
FAIL	github.com/confidential-containers/cloud-api-adaptor/test/e2e	1288.319s
```
It seems like a recent regression as that part was working fine before.

**Comment by wainersm** [comment]
Updated to use cloud-api-adaptor:dev-27bf2ba0730eb9f7eba9923c46eb9b41e771ce08 image (a.k.a the latest) to see if it gets installed as used to be.

**Comment by wainersm** [comment]
Resolved conflict with the image workflow and added debug to tests failing.

**Comment by wainersm** [comment]
And here is another test failing:

```
=== RUN   TestLibvirtCreatePodWithConfigMap/ConfigMapPeerPod_test
    common_suite_test.go:192: Expected Pod State: Running
    common_suite_test.go:193: Current Pod State: Running
=== RUN   TestLibvirtCreatePodWithConfigMap/ConfigMapPeerPod_test/Configmap_is_created_and_contains_data
    common_suite_test.go:277: Command stdout: 
    common_suite_test.go:278: Command stderr: 
    common_suite_test.go:279: unable to upgrade connection: container not found ("nginx-configmap-container")
time="2023-08-23T23:09:12Z" level=info msg="Deleting Configmap... nginx-configmap"
time="2023-08-23T23:09:12Z" level=info msg="Pod logs: /docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration\n/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/\n/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh\n10-listen-on-ipv6-by-default.sh: info: Getting the checksum of /etc/nginx/conf.d/default.conf\ndpkg-query: no packages found matching nginx\n10-listen-on-ipv6-by-default.sh: info: /etc/nginx/conf.d/default.conf differs from the packaged version\n/docker-entrypoint.sh: Sourcing /docker-entrypoint.d/15-local-resolvers.envsh\n/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh\n2023/08/23 23:09:10 [emerg] 1#1: getpwnam(\"nginx\") failed in /etc/nginx/nginx.conf:2\nnginx: [emerg] getpwnam(\"nginx\") failed in /etc/nginx/nginx.conf:2\n/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh\n/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/.wh..wh..opq\n/docker-entrypoint.sh: Configuration complete; ready for start up"
time="2023-08-23T23:09:12Z" level=info msg="Deleting pod... nginx-configmap-pod"
--- FAIL: TestLibvirtCreatePodWithConfigMap (138.90s)
    --- FAIL: TestLibvirtCreatePodWithConfigMap/ConfigMapPeerPod_test (138.90s)
        --- FAIL: TestLibvirtCreatePodWithConfigMap/ConfigMapPeerPod_test/Configmap_is_created_and_contains_data (6.80s)
```

**Comment by wainersm** [comment]
@bpradipt @stevenhorsman I finally made this work, however on code path was not tested: the job that builds the CAA image (although I tested it on a fork) because it should run on pull_request_target (meaning the workflow executed should be in the main branch). I had to disable two tests that are failing only on CI, but I will be investigating the problem once this is merged.

I appreciate if you can make a "loosely" review as I would like to have it merged asap in order to 1) check whether it really works then send fixes; 2) start testing the PRs on this project against the libvirt provider

**Comment by stevenhorsman** [comment]
Do we have any issues like the Azure tests where we need to have an approval before the libvirt tests run to avoid exposing secrets/Fabiano using the nodes for Crypto mining?

**Comment by wainersm** [comment]
> Do we have any issues like the Azure tests where we need to have an approval before the libvirt tests run to avoid exposing secrets/Fabiano using the nodes for Crypto mining?

Good question!

The workflow needs a privileged token to publish the caa images on the ghcr.io. As the workflow run on pull_request_target it is not possible for a pull request to run a malicious workflow, that's the first fence of protection. Second being the label, someone needs to label the PR as 'test_e2e_libvirt' otherwise it won't run. Only committers should be allowed to put labels on PRs, so this is basically the approval mechanism implemented. Am I wrong and anyone can label a PR?

Unlike some workflows on kata, here it is not explicitly instantiating any resources on Azure so we don't even need the secrets configured on the project. The runners are auto-magically  deployed by garm and handed over as self-hosted runners. This is the third security fence, a.k.a, anti-Fabiano's crypto scam ;)

File: .github/workflows/e2e_libvirt.yaml
Comment (stevenhorsman) [inline .github/workflows/e2e_libvirt.yaml:142]: I might come to it in a later commit, but is Teardown no so we can get debug info and the test runner node should be cleaned up anyway at the end of the workflow?

Comment (wainersm) [reply .github/workflows/e2e_libvirt.yaml:142]: Exactly Steve!

**Comment by wainersm** [comment]
> I've given it a non-extensive review and it looks good. I wonder if we can start to re-use pieces of this in the other (e.g. Azure, but maybe internal IBM) tests too, but that's something for when the code is more tested. Thanks for the great work Wainer

Thanks for the review sir!

My aim was to create a reusable workflow indeed, for example, the podvm (that should be improved) and CAA build jobs are meant generic. But I didn't have time to look in deep how the CI for the other providers are implemented today. Definitely I want to have more shareable code among the providers CI.

File: .github/workflows/e2e_on_pull.yaml
Comment (mkulke) [inline .github/workflows/e2e_on_pull.yaml:34]: it's a dummy, but it should be `- run: "true"` otherwise it's interpreted as boolean

File: .github/workflows/e2e_on_pull.yaml
Comment (mkulke) [inline .github/workflows/e2e_on_pull.yaml:120]: Could providers be a matrix?

**Review by mkulke — APPROVED** [review]

File: .github/workflows/e2e_on_pull.yaml
Comment (wainersm) [reply .github/workflows/e2e_on_pull.yaml]: hey @mkulke, first of all, thanks for the review. For this comment in particular, yeah, I hadn't tested that code path...good catch. Fixed.

File: .github/workflows/e2e_on_pull.yaml
Comment (wainersm) [reply .github/workflows/e2e_on_pull.yaml:120]: Yes, they could but I thought it would be overkill. My idea is to have a single `install` directory to be shared by all downstream jobs (i.e. provider specific jobs). If I convert to gha matrix it will create N jobs (N VMs instantiated) to produce the same artifact, it seems just resources waste.

**Comment by wainersm** [comment]
Updated to address the comment of @mkulke regaring `run: true`. Also rebased to main.

File: .github/workflows/e2e_libvirt.yaml
Comment (ldoktor) [inline .github/workflows/e2e_libvirt.yaml:95]: @wainersm Shouldn't this be a required step for `Install kcli` rather than handling ssh-keygen there as well?

File: test/e2e/libvirt_test.go
Comment (ldoktor) [inline test/e2e/libvirt_test.go:18]: How about using an ENV value to skip it only when running in CI while allowing it otherwise?

File: .github/workflows/e2e_libvirt.yaml
Comment (wainersm) [reply .github/workflows/e2e_libvirt.yaml:95]: hi @ldoktor , I'm glad to see you here :)

This key is used by cloud-api-adaptor to connect to the Libvirt on host. It could be the same key as used by the kcli to install the cluster, but I prefer to keep them two separated keys.

**Pull request merged**
